### ISSUE TITLE
eslint does not error when using var instantiated on same line

### ISSUE THREAD

**Issue by DelvarWorld** [eslint/eslint#5359]
I don't think it should be allowed by eslint, even though it's technically valid javascript (I don't think it creates an implicit global val?). This just bit me and it seemed like eslint should have caught it.

Code:
```js
const obj = {};
const val = obj[ val ];
```

```sh
echo 'const obj = {}; const val = obj[ val ];' | eslint --stdin
```

v2.2.0

**Comment by eslintbot** [comment]
Hi @DelvarWorld, thanks for the issue. It looks like there's not enough information for us to know how to help you. If you're reporting a bug, please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
2. What you did (the source code and ESLint configuration)
3. The actual ESLint output complete with numbers
4. What you expected to happen instead

Requesting a new rule? Please see [Proposing a New Rule](http://eslint.org/docs/developer-guide/contributing/new-rules) for instructions.

If it's something else, please just provide as much additional information as possible. Thanks!

**Comment by DelvarWorld** [comment]
I appreciate @eslintbot but the original ticket contains all of that information.

**Comment by ilyavolodin** [comment]
@DelvarWorld It looks like you are proposing new rule. Could you follow our instructions please? I'm not at all clear why you think the pattern that you posted above should error.

**Comment by DelvarWorld** [comment]
Explicitly:

**The version of ESLint you are using (run eslint -v)**
v2.2.0

**What you did (the source code and ESLint configuration)**

```sh
node --version
    v4.2.3
npm install -g eslint
eslint --version
    v2.2.0
echo 'const obj = {}; const val = obj[ val ]];' | eslint --stdin
```

**The actual ESLint output complete with numbers**
There is not output because it passes the eslint check.

**What you expected to happen instead**
You can see on this line:
```js
const val = obj[ val ];
```

That `val` is both the declared variable name and the object key lookup. I (think) this is technically valid Javascript and I (think) it doesn't create a global variable `val` by accident, but this code is clearly not correct to me. I expect eslint to notice that I'm using a variable as a key that hasn't been instantiated yet.

**Comment by mysticatea** [comment]
Can you use `no-use-before-define` rule?

**Comment by DelvarWorld** [comment]
@mysticatea with
```json
"rules": {
    "no-use-before-define": 1
}
```
in my `.eslintrc`, same behavior, no error

**Comment by mysticatea** [comment]
Hmm, it's strange...

In my local:

```
> eslint -v
v2.2.0
> echo 'const obj = {}; const val = obj[ val ];' | eslint --stdin --no-eslintrc --env es6 --rule no-use-before-define:2

<text>
  1:34  error  'val' was used before it was defined  no-use-before-define

✖ 1 problem (1 error, 0 warnings)
```

**Comment by DelvarWorld** [comment]
Ah ok, using the `no-use-before-define` rule **does** work for me, thanks. I was using an older version of eslint by accident by forgetting to select the correct node version with nvm in a new shell.
